### DISCUSSION TITLE
Yalmip slow model building

### DISCUSSION THREAD

**Discussion by TomBloemers** [yalmip/YALMIP#1051] [Optimization modelling] [answered]
Dear YALMIP users,

I am having some issues with yalmip, where the process of building a model (formulating quadratic inequality constraints) is extremely slow. I've already vectorized as much as I can to speed things up. Are there any other tricks to speed up the process, or are there some mistakes that I make? 

Sample code snippit below. I am currently testing this code with cNk and cDk in R^(6x1). Where each constraint line is a 999x1 complex valued array. This already runs incredibly slow, with the 5 constraint lines taking around 15 seconds.

Thanks in advance for the help!

```
N = 1000;
nb = 6;
NkB = randn(N, nb) + 1i*randn(N, nb);
DkB = randn(N, nb) + 1i*randn(N, nb);
Ng = randn(N, 1) + 1i*randn(N, 1);
Dg = randn(N, 1) + 1i*randn(N, 1);
Ws = randn(N, 1) + 1i*randn(N, 1);
Wt = randn(N, 1) + 1i*randn(N, 1);
Wr = randn(N, 1) + 1i*randn(N, 1);
Wp = randn(N, 1) + 1i*randn(N, 1);
p = 1;

% Opt. variables
cNk = sdpvar(nb, 1, 'full');
cDk = sdpvar(nb, 1, 'full');
gam = sdpvar(1);

Nk = NkB*cNk*p;
Dk = DkB*cDk*p;

constr = (cDk(1, :) == [1, zeros(1, 1 - 1)]);
constr = constr + (abs( Dg.*Dk.*Ws ) <= gam*real( Ng.*Nk + Dg.*Dk ));
constr = constr + (abs( Dg.*Nk.*Wr ) <= gam*real( Ng.*Nk + Dg.*Dk ));
constr = constr + (abs( Ng.*Dk.*Wp ) <= gam*real( Ng.*Nk + Dg.*Dk ));
constr = constr + (abs( Ng.*Nk.*Wt ) <= gam*real( Ng.*Nk + Dg.*Dk ));

sdpOpts = sdpsettings('solver', 'mosek');
optobj   = optimizer(constr, [], sdpOpts, {gam}, {cNk cDk});

**Comment by johanlofberg** [comment, marked as answer]
Unclear what is constant and what is decisions. You declare cNk as an output, but never define it as a variable. Supply a reproducible example

**Reply by TomBloemers** [reply]
Sorry, I missed the line where cNk is declared. I've added some dummy variables as the code that generates the variables is quite extensive. The solution to these random variables might not be feasible, but the problem is with the constraint lines. I've updated the code snippit in the original post.

**Reply by johanlofberg** [reply]
15 seconds sounds pretty slow. I have a slow laptop and it takes 7s, but maybe yours is even slower

Anyhow, the issue here is that the abs-operator is an elementwise operator, so it has to work with elementwise modelling of all 4000 expressions, which simply takes time. Pretty much worst-case scenario. Much more efficient to simply model large absolute values abs(x)<=y manually here as -y<=x<=t

**Reply by johanlofberg** [reply]
and now I saw that it is abs on a complex vector, i.e. every single element will require an socp-cone. To get that fast, you will have to model it manually using a vectorized cone operator call which in the lines of cone([all-right-hand_sides;real(x);imag(x)])

**Reply by TomBloemers** [reply]
Thanks! I will try it out and report back.

**Reply by TomBloemers** [reply]
Dear Johan,

Unfortunately, the usage of "cone" as described in your reaction above does not yield a similar result, in fact my original problem becomes infeasible for a solution that I know is most certainly feasible. I've included a comparison of the two methods below. Unfortunately, my knowledge is lacking in this domain, so I am not sure what I am doing wrong here.

```
% Dummy variables
N = 1;
nb = 1;
NkB = randn(N, nb) + 1i*randn(N, nb);
DkB = randn(N, nb) + 1i*randn(N, nb);
Ng = randn(N, 1) + 1i*randn(N, 1);
Dg = randn(N, 1) + 1i*randn(N, 1);
Ws = randn(N, 1) + 1i*randn(N, 1);
Wt = randn(N, 1) + 1i*randn(N, 1);
Wr = randn(N, 1) + 1i*randn(N, 1);
Wp = randn(N, 1) + 1i*randn(N, 1);
p = 1;

% Opt. variables
cNk = sdpvar(nb, 1, 'full');
cDk = sdpvar(nb, 1, 'full');
gam = sdpvar(1);

Nk = NkB*cNk*p;
Dk = DkB*cDk*p;

constr = (cDk(1, :) == [1, zeros(1, 1 - 1)]);
constr = constr + (abs( Dg.*Dk.*Ws ) <= gam*real( Ng.*Nk + Dg.*Dk ));
constr = constr + (abs( Dg.*Nk.*Wr ) <= gam*real( Ng.*Nk + Dg.*Dk ));
constr = constr + (abs( Ng.*Dk.*Wp ) <= gam*real( Ng.*Nk + Dg.*Dk ));
constr = constr + (abs( Ng.*Nk.*Wt ) <= gam*real( Ng.*Nk + Dg.*Dk ));

constr2 = (cDk(1, :) == [1, zeros(1, 1 - 1)]);
constr2 = constr2 + cone([gam*real( Ng.*Nk + Dg.*Dk ); real(Dg.*Dk.*Ws); imag(Dg.*Dk.*Ws)]);
constr2 = constr2 + cone([gam*real( Ng.*Nk + Dg.*Dk ); real(Ng.*Dk.*Wp); imag(Ng.*Dk.*Wp)]);
constr2 = constr2 + cone([gam*real( Ng.*Nk + Dg.*Dk ); real(Dg.*Nk.*Wr); imag(Dg.*Nk.*Wr)]);
constr2 = constr2 + cone([gam*real( Ng.*Nk + Dg.*Dk ); real(Ng.*Nk.*Wt); imag(Ng.*Nk.*Wt)]);

sdpOpts = sdpsettings('solver', 'mosek');
optobj   = optimizer(constr, [], sdpOpts, {gam}, {cNk cDk});
optobj2   = optimizer(constr2, [], sdpOpts, {gam}, {cNk cDk});

**Reply by johanlofberg** [reply]
Looks like you are creating a cone with one looooong vector, instead of a matrix 3xN

**Reply by TomBloemers** [reply]
Thanks for all the help. Correcting the dimensions indeed fixed the problem, which is now significantly faster than before.
